### PR TITLE
Add initial implementation of Query expression

### PR DESCRIPTION
**Pull request by mohanvive**
## Purpose
This PR contains the initial implementation of query expression.

## Approach
N/A

## Samples
N/A

## Remarks
Refer, related issue https://github.com/ballerina-platform/ballerina-lang/issues/20722 ...

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/clauses/FromClauseNode.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ *  under the License.
+ */
+
+package org.ballerinalang.model.clauses;
+
+import org.ballerinalang.model.tree.Node;
+import org.ballerinalang.model.tree.expressions.ExpressionNode;
+import org.ballerinalang.model.tree.statements.VariableDefinitionNode;
+
+/**
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/model/clauses/FromClauseNode.java]: Need to add @since

Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/model/clauses/FromClauseNode.java]: applies to other files as well

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/QueryExpressionNode.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+package org.ballerinalang.model.tree.expressions;
+
+import org.ballerinalang.model.clauses.FromClauseNode;
+import org.ballerinalang.model.clauses.SelectClauseNode;
+import org.ballerinalang.model.clauses.WhereClauseNode;
+
+import java.util.List;
+
+/**
+ * Expression of (expression).
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/QueryExpressionNode.java]: IMO the description is not meaningful

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -4324,6 +4328,139 @@ public void visit(BLangStatementExpression bLangStatementExpression) {
...
+        BLangForeach parentForEach = null;
+
+        for (BLangFromClause bLangFromClause : fromClauseList) {
+            BLangForeach foreach = (BLangForeach) TreeBuilder.createForeachNode();
+            foreach.pos = queryExpr.pos;
+            foreach.collection = bLangFromClause.collection;
+            types.setForeachTypedBindingPatternType(foreach);
+
+            foreach.variableDefinitionNode = bLangFromClause.variableDefinitionNode;
+            foreach.isDeclaredWithVar = true;
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Why do we set this field to true always?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -4324,6 +4328,139 @@ public void visit(BLangStatementExpression bLangStatementExpression) {
...
+                ASTBuilderUtil.createVariableDef(pos, outputArrayVariable);
+        BLangSimpleVarRef outputVarRef = ASTBuilderUtil.createVariableRef(pos, outputArrayVariable.symbol);
+
+        // Create indexed based access expression statement
+        //      x[x.length()] = {
+        //         firstName: person.firstName,
+        //         lastName: person.lastName
+        //      };
+
+        if (selectClause.expression.type == null) {
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: this is always not null after typechecker?

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/clauses/FromClauseNode.java (hunk 1 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/model/clauses/FromClauseNode.java]: Done.

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/QueryExpressionNode.java (hunk 2 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/QueryExpressionNode.java]: Done.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 4 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 5]
```diff
@@ -4324,6 +4328,139 @@ public void visit(BLangStatementExpression bLangStatementExpression) {
...
+            // Create If Statement with Where expression and foreach body
+            BLangIf outerIf = null;
+            BLangIf innerIf = null;
+            for (BLangWhereClause whereClause : whereClauseList) {
+                BLangIf bLangIf = (BLangIf) TreeBuilder.createIfElseStatementNode();
+                bLangIf.pos = queryExpr.pos;
+                bLangIf.expr = whereClause.expression;
+                if (innerIf != null) {
+                    BLangBlockStmt bLangBlockStmt = ASTBuilderUtil.createBlockStmt(pos);
+                    bLangBlockStmt.addStatement(bLangIf);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]:
```suggestion
                    bLangBlockStmt.stmts.add(bLangIf);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 3 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 6]
```diff
@@ -4324,6 +4328,139 @@ public void visit(BLangStatementExpression bLangStatementExpression) {
...
+
+        BLangInvocation lengthInvocation = createLengthInvocation(pos, outputArrayVariable.symbol);
+        lengthInvocation.expr = outputVarRef;
+        BLangIndexBasedAccess indexAccessExpr = ASTBuilderUtil.createIndexAccessExpr(outputVarRef, lengthInvocation);
+        indexAccessExpr.type = selectClause.expression.type;
+
+        BLangAssignment outputVarAssignment = ASTBuilderUtil.createAssignmentStmt(pos, indexAccessExpr,
+                selectClause.expression);
+        // Set the indexed based access expression statement as foreach body
+        foreachBody.addStatement(outputVarAssignment);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]:
```suggestion
        foreachBody.stmts.add(outputVarAssignment);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 7]
```diff
@@ -4324,6 +4328,139 @@ public void visit(BLangStatementExpression bLangStatementExpression) {
...
+            BLangIf outerIf = null;
+            BLangIf innerIf = null;
+            for (BLangWhereClause whereClause : whereClauseList) {
+                BLangIf bLangIf = (BLangIf) TreeBuilder.createIfElseStatementNode();
+                bLangIf.pos = queryExpr.pos;
+                bLangIf.expr = whereClause.expression;
+                if (innerIf != null) {
+                    BLangBlockStmt bLangBlockStmt = ASTBuilderUtil.createBlockStmt(pos);
+                    bLangBlockStmt.addStatement(bLangIf);
+                    innerIf.setBody(bLangBlockStmt);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]:
```suggestion
                    innerIf.body = bLangBlockStmt;
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 4 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Now, record literals are not type checked hence we need this check now. We can remove this once lang level changes are done.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 8]
```diff
@@ -4324,6 +4328,139 @@ public void visit(BLangStatementExpression bLangStatementExpression) {
...
+                } else {
+                    outerIf = bLangIf;
+                }
+                innerIf = bLangIf;
+            }
+            innerIf.setBody(foreachBody);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]:
```suggestion
            innerIf.body = foreachBody;
```

**Pull request merged**